Add utility script to batch filter images

carshadi opened the pull request:
This script performs Frangi Vesselness filtering on all image files in a directory (with optional recursion).

Rationale:
Tracing on filtered images is one of SNT's most powerful features. However, it is currently only possible to generate one filtered image at a time from within SNT. This script would enable batch pre-processing of images directly from SNT.

Tested on Windows 10 64-bit and Windows 7 64-bit.

carshadi committed:
Add utility script to batch filter images

This script performs Frangi Vesselness filtering on all TIFF images in a directory (with optional recursion)

carshadi committed:
Add checks to verify 2D/3D grayscale

add check to verify 2D/3D grayscale image
add check that scale parameter is >= 0
improve logging

tferr commented on src/main/resources/script_templates/Neuroanatomy/Utilities/Filter_Multiple_Images.py:
```diff
@@ -0,0 +1,97 @@
+
+        # Create placeholder image for the output then run the Frangi Vesselness op
+        output = ij.op().run("create.img", float_input)
+        ij.op().run("frangiVesselness", output, float_input, spacing, scale)
```
@carshadi, I had the impression that the Frangi op expects the `scale` parameter to be in pixel units rather than spatially calibrated units. Is that the case? (pinging @gselzer)

gselzer replied:
Yes, @tferr, this is correct.

carshadi replied:
i see, so would the fix simply be to change ```# @float(label="Size of structures to be filtered (spatially calibrated units)") scale``` to ```# @float(label="Size of structures to be filtered (pixel units)") scale```? Or would the spacing need to be changed as well?

gselzer replied:
`spacing` is used by the Op to refer to the physical distance between each sample, which you seem to calculate right. I also think that change to pixel units is sufficient.

tferr replied:
(Hopefully, you guys will not think I'm being anal, but I disagree).
it is _way_ more convenient to work on physical units. Take the following example for raster confocal microscopy: One acquires several images each containing a single cell (e.g., a neuron). Because cell size varies, one tweaks digital zoom so that a cell is compliteley included in field of view. The zoom adjustments respect the Nyquist criterion and are valid. One places all those images in a directory and tries to run this script. How would that work? Split the images by pixel size? "Eye-ball" a common value? I sincerely think, most users know the size of the features they want to enhance (e.g., these dendrites range between 6-10um, a soma  is ~10um, etc), but they may have no idea what to the _exact_  pixels that corresponds to. That info is stored in the image metadata. Furthermore, real-world coordinates are transferable across microscopy modalities and allow scripts to run as-is across labs. So, in short, this is what I would propose:
1) Keep the prompt in physical units
2) Compute the pixel size as the average of the voxel dimensions (one needs to take care that with 2D images, any existing metadata on depth (Z-dimension) is not averaged). Reducing voxel dimensions to this averaged isotropic dimension is a common, reasonable assumption.
3) Add a description to the prompt that this is what happening.

carshadi replied:
I'm not sure I follow. Compute the ```scale``` parameter as the average of the voxel dimensions?

carshadi replied:
or do that with the spacing and force it to be isotropic?

tferr replied:
The former:

```python
scale /= (sum(spacing) / len(spacing))  ## average voxel size: convert scale to pixel units
ij.op().run("frangiVesselness", output, float_input, spacing, scale)
```

carshadi replied:
So then remove the input from the prompt and just say that's what's happening? Or allow them to input a multiplier for the later calculated "baseline"?

carshadi replied:
oh i missed the /=, i see will fix it

carshadi committed:
Add conversion from physical units to pixel units

-convert scale parameter to pixel units using average voxel size
- add a check that third axis is depth

carshadi committed:
Compare against Axes.Z instead of string label

carshadi commented:
changed

tferr commented:
Merged! Thanks a lot! Just added some minor tweaks in 2a0a0219a6be63ed631afd6b605fef7596814d41 so that the prompt does not get too wide
